### PR TITLE
Add support for gjs/gts files

### PR DESCRIPTION
**Pull request by vstefanovic97**
> **Disclaimer:** Disclaimer this is just to make the js/ts part of the code have coverage, it does not support covering hbs part of the code. this is done in https://github.com/ember-cli-code-coverage/ember-cli-code-coverage/pull/319

So to sum it up the steps in this PR are:

1. Add support for inputSourceMaps (remap coverage using https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-source-maps)

When working with already transpiled code, that has sourcemaps included (inline or in separate file) in order for the lines to be correct we actually need to use [istanbul-lib-source-maps](https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-source-maps) to remap the coverage to the correct source code.

This is actually needed since when we build v2 addons we already transpile the code from `gjs/gts` to `ts/js` first and we run babel on the already transpiled code, so we need to remap the coverage to the original source code.
Disclaimer: gts/gjs->js/ts transpilation doesn't generate sourcemaps at the moment but there is a PR open which will enable it https://github.com/embroider-build/content-tag/pull/62

2. Make sure to ignore template's `eval` method from the coverage report

Because we instrument already transpiled code we end up instrumenting some method that shouldn't be instrumented, like for example

```gjs
const var baz = <template>
  <button {{on "click" bar}}>
   increment
  </button>
</template>
```

gets instrumented to:

```js
var baz = template(`
  <button {{on "click" bar}}>
   increment
  </button>
`, {
  eval() {
    cov_1sbgmdzd2v().f[2]++; // we don't want istanbul to track this
    cov_1sbgmdzd2v().s[5]++; // we don't want istanbul to track this
    return eval(arguments[0]);
  }
});
```
Because istanbul instruments the `eval` method it will report as uncovered in tests and also it skews coverage percentages, so we want it to not instrument this, for this I created a small babel plugin that will add istanbul ignore lines and run before istanbuls babel plugin.

3. Include `.gjs/.gts` extensions in istanbul `extension` option by default so instrumentation will work on these files

4. Write tests (create a v2 addon with gjs/gts where we build it with coverage and then tests the coverage with snapshots)
  Lines in the snapshots coverage will still be offset until https://github.com/embroider-build/content-tag/pull/62 is fixed and 
  we update the test app with the change

### CODE REVIEW

**Comment by vstefanovic97** [comment]
@kategengler seems like the error reported in CI was a known PNPM issue https://github.com/pnpm/pnpm/issues/5056 which was fixed with the latest pnpm version. I updated the version for CI and Volta do you mind to rerun the workflow and see if it fixes it?

**Comment by vstefanovic97** [comment]
@RobbieTheWagner can you or maybe some other maintainer please review this?

File: test-packages/my-embroider-app-test.mjs [hunk 1]
```diff
@@ -22,6 +22,7 @@ describe('app coverage generation', function () {
         "@embroider/webpack": embroiderVersion
       });
   
+      await execa('rm',  ['-rf', '.embroider'], { cwd: buildPath, env });
```
Comment (RobbieTheWagner) [inline test-packages/my-embroider-app-test.mjs:25]: Can you please explain why this was needed?

Comment (vstefanovic97) [reply test-packages/my-embroider-app-test.mjs:25]: @RobbieTheWagner  so currently we are testing against `emroider/core` versions from 0 to 3.
During each build it creates a `.embroider` folder, but seems like when we run a test for a newer version it will try to re-use the `.embrodier` folder generated by the previous test and it was leading to some unpredictability.

For example there is a problem with [babel plugin serialization](https://github.com/embroider-build/embroider/issues/1525) in v3 that I solved [here](https://github.com/ember-cli-code-coverage/ember-cli-code-coverage/pull/405/files#diff-220612c243d58fc738e2512fd4efbd91e7700e2b8b547032a3e260df5fb7ee9dR92) using [this](https://github.com/Serabe/ember-bind-helper/pull/21) as a reference, but it would still fail on tests w/o cleaning up `.embroider` because it was relying on what was generated by the previous embroider version

So basically all I did was make sure the tests are totally isolated from each other and don't affect each other

Comment (SergeAstapov) [reply test-packages/my-embroider-app-test.mjs:25]: @vstefanovic97 that's good knowledge! Mind to drop a comment in the code with this explanation for future readers?

Comment (vstefanovic97) [reply test-packages/my-embroider-app-test.mjs:28]: Added a comment in the code about this @SergeAstapov, @RobbieTheWagner

**Comment by vstefanovic97** [comment]
@RobbieTheWagner sorry to bother you again, any progress with the review?
Is there anything I can do to help make the reviewing easier, or anything additional that need explaining, any concerns etc. ?

**Comment by RobbieTheWagner** [comment]
@vstefanovic97 I really don't have enough knowledge about how this addon works to review. @SergeAstapov do you feel comfortable reviewing?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -20,7 +20,7 @@ jobs:
       - uses: pnpm/action-setup@v2
         name: Install pnpm
         with:
-          version: 8.10.3
+          version: 8.15.1
```
Comment (SergeAstapov) [inline .github/workflows/ci.yml:23]: @vstefanovic97 is this necessary for this PR? Can we do this bump separately?

File: packages/ember-cli-code-coverage/index.js [hunk 3]
```diff
@@ -48,19 +58,25 @@ module.exports = {
       if (config.coverageEnvVar) {
         coverageEnvVar = config.coverageEnvVar;
       }
+
+      if (config.extension) {
```
Comment (SergeAstapov) [inline packages/ember-cli-code-coverage/index.js:62]: @vstefanovic97 what is the use case when we want to configure this?
Also, as this exposes new option for consumers, let's document it in readme

File: packages/ember-cli-code-coverage/index.js [hunk 4]
```diff
@@ -72,7 +88,12 @@ module.exports = {
     }
 
     const IstanbulPlugin = require.resolve('babel-plugin-istanbul');
-    return [[IstanbulPlugin, { cwd, include: '**/*', exclude }]];
+    return [
+      useStringLookupForPlugin
```
Comment (SergeAstapov) [inline packages/ember-cli-code-coverage/index.js:92]: @vstefanovic97 could you please explain this ternary? can't we do e.g. `require` all the time?

File: package.json [hunk 5]
```diff
@@ -8,6 +8,7 @@
     "test": "vitest"
   },
   "devDependencies": {
+    "@babel/core": "^7.23.9",
```
Comment (SergeAstapov) [inline package.json:11]: @vstefanovic97 why is this needed in the root of monorepo? I guess each package that needs it should declare such kind of dependency.

File: packages/ember-cli-code-coverage/lib/attach-middleware.js [hunk 6]
```diff
@@ -134,17 +137,22 @@ function adjustCoverage(coverage, options) {
       namespaceMappings,
       modifyAssetLocation
     );
-    coverage[filePath].path = path.relative(root, relativeToProjectRoot);
-    memo[path.relative(root, relativeToProjectRoot)] = coverage[filePath];
+    coverage[filePath].data.path = path.relative(root, relativeToProjectRoot);
```
Comment (SergeAstapov) [inline packages/ember-cli-code-coverage/lib/attach-middleware.js:140]: @vstefanovic97 out of curiosity, is this change needed because we now work with struct returned by `sourceMapStore.transformCoverage`?

File: packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    visitor: {
+      Program: {
+        enter(path, state) {
+          const filename = state.file.opts.filename;
+
+          if (!filename?.match(/\.g[tj]s$/)) {
+            return;
+          }
+
+          path.traverse(gjsGtsTemplateIgnoreVisitor, state); // we need to do early traverse to make sure this runs before istanbuls plugin
```
Comment (SergeAstapov) [inline packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js:42]:
```suggestion
          // we need to do early traverse to make sure this runs before istanbuls plugin
          path.traverse(gjsGtsTemplateIgnoreVisitor, state);
```
just a tiny bit more readable

File: packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
+const gjsGtsTemplateIgnoreVisitor = {
```
Comment (SergeAstapov) [inline packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js:1]: @vstefanovic97 let's provide explanation why we add this, which problem it solves and which other options e have considered but didn't go and why.

File: test-packages/my-v2-addon-gjs-gts/addon/babel.config.cjs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    require('decorator-transforms'),
+    [
+      'babel-plugin-ember-template-compilation',
+      {
+        targetFormat: 'hbs',
+        transforms: [],
+      },
+    ],
+    ...require('ember-cli-code-coverage').buildBabelPlugin({
+      extension: ['.gjs', '.gts', '.js', '.ts', '.mts', '.cts', '.cjs', '.mjs'],
```
Comment (SergeAstapov) [inline test-packages/my-v2-addon-gjs-gts/addon/babel.config.cjs:22]: @vstefanovic97 this matches the default value so we can probably skip this?

File: test-packages/my-v2-addon-gjs-gts/addon/rollup.config.mjs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    babel({
+      extensions: ['.js', '.gjs', '.ts', '.gts'],
+      babelHelpers: 'bundled',
+    }),
+
+    // Ensure that standalone .hbs files are properly integrated as Javascript.
+    addon.hbs(),
+
+    // Ensure that .gjs files are properly integrated as Javascript
+    addon.gjs({ inline_source_map: true }),
```
Comment (SergeAstapov) [inline test-packages/my-v2-addon-gjs-gts/addon/rollup.config.mjs:55]: @vstefanovic97 looks like it's worth to document this option in readme.
Maybe, I'm thinking we need a section with steps how to setup test coverage in v2 addons.

**Review by SergeAstapov — CHANGES_REQUESTED** [review]
@RobbieTheWagner Thank you for pinging me!
Functionally, this looks great to me! I'd suggest we spend a bit more effort on documentation part to make clear for ppl who'd like to start using this and for ppl who would need to understand code in future

File: packages/ember-cli-code-coverage/index.js (hunk 3 above)
Comment (vstefanovic97) [reply packages/ember-cli-code-coverage/index.js:62]: @SergeAstapov I can't think of a use case right now, but thought maybe it would be useful to expose it still, I'll add it to the docs

File: packages/ember-cli-code-coverage/index.js (hunk 4 above)
Comment (vstefanovic97) [reply packages/ember-cli-code-coverage/index.js:92]: There is a problem with [babel plugin serialization](https://github.com/embroider-build/embroider/issues/1525) in v3, I solved it using string lookups since I found [this](https://github.com/Serabe/ember-bind-helper/pull/21) as a reference.
I tried to use string lookup all the time, but it didn't work for v1 and v2

File: package.json (hunk 5 above)
Comment (vstefanovic97) [reply package.json:11]: It's used in tests [here](https://github.com/ember-cli-code-coverage/ember-cli-code-coverage/pull/405/files#diff-11851204535d7735df82299a725e271fc96f8c040a5468c6070cba4a7ce479f6R2)

File: packages/ember-cli-code-coverage/lib/attach-middleware.js (hunk 6 above)
Comment (vstefanovic97) [reply packages/ember-cli-code-coverage/lib/attach-middleware.js:140]: yes @SergeAstapov

File: .github/workflows/ci.yml (hunk 2 above)
Comment (vstefanovic97) [reply .github/workflows/ci.yml:23]: @SergeAstapov it's necessary, see https://github.com/ember-cli-code-coverage/ember-cli-code-coverage/pull/405#issuecomment-1916394358

File: package.json (hunk 5 above)
Comment (SergeAstapov) [reply package.json:11]: ah, now I see - `test-packages` folder contains both vite tests *and* test packages.
IMO seems confusing but for sure we should follow existing structure in this PR.

File: packages/ember-cli-code-coverage/index.js (hunk 3 above)
Comment (vstefanovic97) [reply packages/ember-cli-code-coverage/index.js:62]: @SergeAstapov I added this option to the docs

File: packages/ember-cli-code-coverage/index.js (hunk 4 above)
Comment (vstefanovic97) [reply packages/ember-cli-code-coverage/index.js:93]: @SergeAstapov Added comment about this to the code

File: packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js (hunk 7 above)
Comment (vstefanovic97) [reply packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js]: Fixed @SergeAstapov

File: packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js (hunk 8 above)
Comment (vstefanovic97) [reply packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin.js:18]: @SergeAstapov I added an explanation to the file

**Comment by vstefanovic97** [comment]
@RobbieTheWagner looks like I had deleted something by accident, I fixed it now, mind running again? This time I think it will passf

File: README.md [hunk 11]
```diff
@@ -154,6 +154,8 @@ Configuration is optional. It should be put in a file at `config/coverage.js` (`
 
 - `excludes`: Defaults to `['*/mirage/**/*']`. An array of globs to exclude from instrumentation. Useful to exclude files from coverage statistics.
 
+- `extension`: Defaults to `['.gjs', '.gts', '.js', '.ts', '.cjs', '.mjs', '.mts', '.cts']`. Tell Istanbul to instrument only files with the provided extensions.
```
[the remark this answers is not in the document]
Comment (vstefanovic97) [reply README.md:157]: @SergeAstapov well in the code it really is `extension` reason for this is I decided to keep it the same as the option in Isntabul since it too is singular

**Comment by vstefanovic97** [comment]
@RobbieTheWagner I fixed latest comments from @SergeAstapov , mind to run CI again, neither I nor @SergeAstapov have permissions to run the CI :/

**Review by SergeAstapov — APPROVED** [review]
if you ask me, this looks great! Thank you @vstefanovic97 for working on this! cc @RobbieTheWagner @kategengler

**Review by RobbieTheWagner — APPROVED** [review]
If @SergeAstapov says this looks good, that is good enough for me 👍

**Comment by vstefanovic97** [comment]
@RobbieTheWagner mind to run again (I think this will be the final time)? I changed test name but forgot to update the snapshot

**Review by SergeAstapov — APPROVED** [review]

**Comment by vstefanovic97** [comment]
@RobbieTheWagner, @SergeAstapov can we merge this and create a new minor release?

**Comment by SergeAstapov** [comment]
IMO yes, we are good to go, but I don't have permissions here.
@RobbieTheWagner would you be able to push this over the line and do the release?

**Comment by cah-brian-gantzler** [comment]
After updating to this version I am now getting no coverage directory being created at all. Anything else I might have to do to configure? I did already have all the configurations from the readme in place for the prior version.

**Comment by vstefanovic97** [comment]
@cah-brian-gantzler there should be no additional config required for it to work.

Do you see any Errors or something?
Is it a regular Ember app/addon, is it Embroider App or v2 addon?

Also if you can come up with a minimal reproduction I could also try to debug it

**Comment by cah-brian-gantzler** [comment]
was on 2.0.3
No do not see any errors. 
it is an embroider app

Since its in my app, doubt and will prob work with a new app, not sure where to start on a minimal not knowing what the problem is.  

Will experiment some. Its weird that it produces NO coverage at all though. Will also revert back and make sure it still produces then. Did last time I asked, but that was a couple weeks ago, was waiting for this update

**Comment by cah-brian-gantzler** [comment]
@vstefanovic97 must have been a glitch. Now produces COVERAGE directory, however no .gjs or .gts files appears appear (same as before)

I checked to ensure I had the correct package in my `node_modules`, and the package.json says `2.1.1` However noticed that the `changelog.md` is out of date. You might want to consider copying the `changelog.md` from the root to the package before publishing. 

I found I also picked up ` useBabelInstrumenter: true,` in the config file over the years. Not sure where I got that, but removing it did not alter the output.

Lastly I have my `mirage` directory in the `app` folder and not the root. Altering the excludes to `*/app/mirage/**/*` or `app/mirage/**/*` did not remove mirage from the report, but `**/mirage/**/*` did although that seems overkill. Got a suggestion on what the glob should be? Not great at globs

**Comment by vstefanovic97** [comment]
@cah-brian-gantzler can you build the app with `COVERAGE=true` it can be either a production or development build, and check if the code for your gjs/gts files is instrumented with Istanbul?

Here is an example of what istanbul should do to the code, if you aren't familiar
https://gist.github.com/robertknight/834452c3b06963ff2a8b9682fd4189cb

Maybe also it's worth checking if you have source maps turned on for your tests
https://github.com/ember-template-imports/ember-template-imports?tab=readme-ov-file#sourcemap-generation
This part is actually needed for the lines to be accurate.

Let's first check this then we can narrow it down further.
I don't have time today, but I'll try this out on a new embroider app to see if it works.

Also I got some feedback here that it works https://github.com/ember-cli-code-coverage/ember-cli-code-coverage/issues/392#issuecomment-1959684127 so maybe what you are seeing is some edge case, but we will try to figure it out and fix :D

**Comment by vstefanovic97** [comment]
@cah-brian-gantzler this looks to be an embroider + ember-cli-code-coverage with gjs/gts issue

I was able to reproduce with a band new embroider app.
I am still trying to dig into the root cause and give it a proper fix

For now can you try to add
```
const path = require('path');

module.exports = {
  modifyAssetLocation(root, relativePath) {
    if (relativePath.includes('/my-embroider-test/')) {
      return path.join('app', relativePath.split('/my-embroider-test/')[1]);
    }

    return false;
  },
};
```
to your `config/coverage.js` file
I think this will get the files to show, but for some reason coverage percentage is still not accurate (I'm still debugging this).
Btw do you mind opening an issue for this, it will be simpler to track there and not in this PR as it's closed

File: packages/ember-cli-code-coverage/index.js [hunk 12]
```diff
@@ -72,7 +88,13 @@ module.exports = {
     }
 
     const IstanbulPlugin = require.resolve('babel-plugin-istanbul');
-    return [[IstanbulPlugin, { cwd, include: '**/*', exclude }]];
+    return [
+      // String lookup is needed to workaround https://github.com/embroider-build/embroider/issues/1525
```
Comment (NullVoxPopuli) [inline packages/ember-cli-code-coverage/index.js:92]: As it turns out... We're going to require string lookup eventually, and forbid passing JS references.

JS references are not parallizable, but string references are! (it's a babel thing :shrug: )

Comment (NullVoxPopuli) [reply packages/ember-cli-code-coverage/index.js:92]: one thing you could try is using a `resolve`d string, so you get the absolute path of the plugin for "this instance of ember-cli-code-coverage".

```js
path.resolve('the-string-you-have'); // returns absolute string, so no matter who is requiring, their resolve isn't local to who is doing the requiring 
```
